### ISSUE TITLE
Pausing the world when the user runs 'cargo build' etc. on the project

### ISSUE THREAD

**Issue by laundmo** [rust-lang/rust-analyzer#17545]
Inspired by 
- #16239  
I would appreciate it if there was a option to automatically pause rust-analyzer when a build process is run, to prevent "waiting for file lock on build directory" or competing for CPU if a separate build directory is set.

This would allow a user to choose to prioritise building the application/example over r-a, which is helpful for faster iteration times.

idea inspired by @vcokltfre

**Comment by Veykril** [comment]
How would this work? The moment you run a cargo command yourself and you get "waiting for file lock on build directory" due to rust-analyzer running something its already too late, since r-a is in fact already running cargo. And r-a can't possibly tell that you want to run cargo yourself at this very moment.

**Comment by redforks** [comment]
I think enable rust-analyzer cargo.targetDir option let rust-analyzer use different build directory can solve the problem. See rust-analyzer User Manual.

rust-analyzer use a different build directory, normal cargo activity including build won't lock because of rust-analyzer.
The drawback is extra hard disk and cpu usage, also memory of course. But rust develop always require a beast machine anyway.

**Comment by laundmo** [comment]
> And r-a can't possibly tell that you want to run cargo yourself at this very moment.

can it not? i was hoping there was a way, maybe specifically watching for cargo processes or something less hacky

> I think enable rust-analyzer cargo.targetDir option let rust-analyzer use different build directory can solve the problem. See rust-analyzer User Manual.

i am aware of this, i even mentioned it in the issue.

**Comment by Veykril** [comment]
> can it not? i was hoping there was a way, maybe specifically watching for cargo processes or something less hacky

Not it can't, watching processes is also not an option I'd consider. I'd recommend disabling check-on-save behavior honestly (and I'l likely push for disabling it by default in the future).

**Comment by flodiebold** [comment]
> I'l likely push for disabling it by default in the future

With the landing page, we could maybe give the user the option more directly :)

**Comment by laundmo** [comment]
hmm but if pausing is implemented/working for #16239 there could be a way for the IDE integration to pause r-a, which it could then do if a build is run through the IDE tooling/debugger?

**Comment by chenyukang** [comment]
> I think enable rust-analyzer cargo.targetDir option let rust-analyzer use different build directory can solve the problem. See rust-analyzer User Manual.
> 
> rust-analyzer use a different build directory, normal cargo activity including build won't lock because of rust-analyzer. The drawback is extra hard disk and cpu usage, also memory of course. But rust develop always require a beast machine anyway.

yes, `cargo.targetDir` resolves this issue a lot, can we set it to true by default?

**Comment by Veykril** [comment]
No, the reason we don't is that it duplicates build artifacts, that is it duplicates disk space used. We should do this instead

> > I'l likely push for disabling it by default in the future
> 
> With the landing page, we could maybe give the user the option more directly :)

That is disabling check on save by default, talking about that feature and its downsides on the landing page and giving a button or something to enable it there. The main problem is we cant really just disable it as people will start scratching their head as to why it no longer works. We probably also need a changelog page first.
